File: key/scripts/tools/checkstyle/publishAudit.py
#!/usr/bin/python3

import sys, os, re, urllib

# workflow:
#  - runIncrementalCheckstyle.sh | tee report.txt
#  - publishAudit.py report.txt

# The original was written in perl and is available under publishAudit.pl

# Written by Alexander Weigl
# Refactored in May 2018, by Mattias Ulbrich

def getenv(*args):
    "relaxed access to environment"
    return map(lambda x: os.environ.get(x, ""), args)

rawReport = sys.argv[1]
SERVER = "https://git.key-project.org/"
URL, PID, TOKEN, SHA, BID, JID = getenv("CI_PROJECT_URL", "CI_PROJECT_ID",
                                        "CI_COMMENT_TOKEN", "CI_COMMIT_SHA",
                                        "CI_BUILD_ID", "CI_JOB_ID")

everythingIsFine = True

with open(rawReport) as raw:
    report = {'ERROR':0, 'WARN':0, 'INFO':0}

    regex = re.compile(r'\[(?P<level>.*?)\] (?P<path>.*\/(?P<file>.*?)):(?P<line>\d+)(:\d+)?: (?P<msg>.*)')

    for line in raw:
        m = regex.match(line)
        if m:
            everythingIsFine = False
            old = report[m.group('level')]
            report[m.group('level')] = old + 1

def r2s(header, count, red=False):
    "report to string"
    if count > 0:
        markup = "**" if red else ""
        return "- " + markup + str(count) + " " + header + markup + "\n"
    else:
        return ""

if everythingIsFine:
    note = """Checkstyle has been run on this commit in [job %s](%s/builds/%s).
    *No issues. Good job*
    """ % (BID, URL, BID)
else:
    note = """Checkstyle has been run on this commit in [job %s](%s/builds/%s).

It found the following issues:
%s

Its report can be [downloaded here](%s/-/jobs/%s/artifacts/raw/report.txt).  
*Please* resolve as many issues as possible before merging the code back into the master branch.  
You can produce a report locally by executing `key/key/scripts/tools/checkstyle/runIncrementalCheckstyle.sh` in your local repository
""" % (BID, URL, BID,
       r2s("ERROR messages", report['ERROR'], red=True) + 
       r2s("warning messages", report['WARN']) +
       r2s("info messages", report['INFO']),
       URL, BID)

print(note)

import requests
reportUrl = "%s/api/v4/projects/%s/repository/commits/%s/comments" % (SERVER,PID,SHA)
print("Send report to", reportUrl)
resp = requests.post(reportUrl, data={ b'private_token': TOKEN, b'note':note })
print(note)
sys.exit(int(bool(report["ERROR"])))

File: scripts/tools/checkstyle/publishAudit.pl
#!/usr/bin/perl -w


#
# THIS PROGRAM IS OUTDATED.
# 
# Please refer to the python program publishAudit.py
# composed by Alexander Weigl.
#
#



# workflow:
#  - runIncrementalCheckstyle.sh | tee report.txt
#  - publishAudit.pl report.txt

use LWP::UserAgent;
use Data::Dumper;
use HTTP::Request::Common;

$rawReport = shift @ARGV;
$server = "https://git.key-project.org/";
$url = $ENV{"CI_PROJECT_URL"};
$pid = $ENV{"CI_PROJECT_ID"};
$token = $ENV{"CI_COMMENT_TOKEN"};
$sha = $ENV{"CI_COMMIT_SHA"};
$bid = $ENV{"CI_BUILD_ID"};

open(my $raw, "<", $rawReport);
my $URL = "hhh";
my %report = ();
while(<$raw>) {

    if(/\[(.*?)\] (.*\/(.*?)):(\d+)(?::\d+)?: (.*)/) {
	$report{$1} .=  "* [$3:$4](../blob/$sha/$2#L$4): $5\n";
    }
}
close($raw);

sub report {
    (my $type, my $msg) = @_;

    my $result = "";
    if($msg) {
        $result .= "#" unless $type eq "ERROR"; # Errors have a larger heading ...
        $result .=  "## $type messages\n\n";
        $result .= $msg . "\n";
    }
    return $result;
}

my $note = "Checkstyle has been run on this commit in [job $bid]($url/builds/$bid).  ";
$note .= "Here is its report:\n\n";
if(%report) {
    $note .= "<details>";
    $note .= &report("ERROR", $report{"ERROR"});
    $note .= &report("WARNING", $report{"WARN"});
    $note .= &report("INFO", $report{"INFO"});
    $note .= "</details>";
} else {
    $note .= "*No issues. Good.*";
}

my $ua = LWP::UserAgent->new();
my $response = $ua->post(
    "$server/api/v4/projects/$pid/repository/commits/$sha/comments",
    { private_token => $token, note => $note });
my $content = $response->as_string();

print Dumper($content);

if($report{"ERROR"}) {
    exit 1;
} else {
    exit 0;
}
